### ISSUE TITLE
ICE `adt::represent_type called on non-ADT type`

### ISSUE THREAD

**Issue by petevine** [rust-lang/rust#30158]
Trying to build something big (rust-snake) with my new rustc 1.6 (i686-unknown-linux-gnu from 2015-11-29, generating code for pentium2) ended on the rocks (pun intended):

````
Compiling image v0.3.11
/home/petevine/.cargo/registry/src/github.com-48ad6e4054423464/image-0.3.11/./src/image.rs:67:13: 67:42 warning: `ImageError::NotEnoughData` does not name a tuple variant or a tuple struct [E0164]
/home/petevine/.cargo/registry/src/github.com-48ad6e4054423464/image-0.3.11/./src/image.rs:67             ImageError::NotEnoughData(..) => &"Not enough data",
                                                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/petevine/.cargo/registry/src/github.com-48ad6e4054423464/image-0.3.11/./src/image.rs:67:13: 67:42 help: run `rustc --explain E0164` to see a detailed explanation
/home/petevine/.cargo/registry/src/github.com-48ad6e4054423464/image-0.3.11/./src/image.rs:67:13: 67:42 note: this warning will become a HARD ERROR in a future release. See RFC 218 for details.
/home/petevine/.cargo/registry/src/github.com-48ad6e4054423464/image-0.3.11/./src/image.rs:67             ImageError::NotEnoughData(..) => &"Not enough data",
                                                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: internal compiler error: adt::represent_type called on non-ADT type: [type error]
note: the compiler unexpectedly panicked. this is a bug.
thread 'rustc' panicked at 'Box<Any>', src/libsyntax/diagnostic.rs:253
stack backtrace:
   1: 0xb72f9750 - sys::backtrace::tracing::imp::write::hbca3a72281dd2320NXt
   2: 0xb7300bd1 - panicking::log_panic::_<closure>::closure.40957
   3: 0xb7300626 - panicking::log_panic::h3bede16486f9aa66m3x
   4: 0xb72c4340 - sys_common::unwind::begin_unwind_inner::hd75e031931020a84kQs
   5: 0xb4174629 - sys_common::unwind::begin_unwind::begin_unwind::h9206301507329329072
   6: 0xb417597b - diagnostic::_<impl>::bug::h35b94f42d9d6620dxQB
   7: 0xb61c45aa - session::_<impl>::bug::h19bd8a985ff2fb96Pll
   8: 0xb6b9dff5 - trans::adt::represent_type_uncached::h1a74a40b34b7fba0mqe
   9: 0xb6b9aa85 - trans::adt::represent_type::h750970da91d0b9aewie
  10: 0xb6cdef92 - trans::_match::compile_submatch_continue::h328c7b8955fdc383HGK
  11: 0xb6cdc985 - trans::_match::compile_submatch::h248d43df6b6b51be5AK
  12: 0xb6ce57a1 - trans::_match::trans_match_inner::hd40a5ce43b2319908aL
  13: 0xb6ca2003 - trans::expr::trans_rvalue_dps_unadjusted::h675335614bd5508cbJC
  14: 0xb6be55cf - trans::expr::trans_into::h5b0ad79c87d8611d7AB
  15: 0xb6be173e - trans::controlflow::trans_block::h6da04dad1703d68cHmw
  16: 0xb6bded70 - trans::base::trans_closure::h0a024109307fe6d90Hi
  17: 0xb6be1cad - trans::base::trans_fn::h347c5a7efc4a77abMRi
  18: 0xb6be776b - trans::base::trans_item::h354de8740df7416aGfj
  19: 0xb6bff7ba - intravisit::walk_item::walk_item::h4569786521516983045
  20: 0xb6bf4075 - trans::base::trans_crate::h28722b15bdd6d4e1mZj
  21: 0xb75ce7b9 - driver::phase_4_translate_to_llvm::h71a5a29c082d906akTa
  22: 0xb75c914f - driver::phase_3_run_analysis_passes::_<closure>::closure.25415
  23: 0xb75a8366 - middle::ty::context::_<impl>::create_and_enter::create_and_enter::h11950113070383832340
  24: 0xb75a3c26 - driver::phase_3_run_analysis_passes::h2556969371712861103
  25: 0xb75781b8 - driver::compile_input::hadceb600e811eacbjca
  26: 0xb756aa57 - run_compiler::hc0bfe5e56a0a613f0wc
  27: 0xb75670d7 - sys_common::unwind::try::try_fn::try_fn::h9525110745468537403
  28: 0xb72ee378 - sys_common::unwind::try::inner_try::hfb6fc30c57459f3aSMs
  29: 0xb7567478 - boxed::_<impl>::call_box::call_box::h2977861881415846031
  30: 0xb72fef39 - sys::thread::_<impl>::new::thread_start::h17b9d21ece202d95Y5w
  31: 0xb3fc7678 - start_thread
  32: 0xb718e58d - clone
  33:        0x0 - <unknown>
````

I was using `cargo-bake` to compile.

**Comment by jdm** [comment]
Should be fixed by #30116.

**Comment by petevine** [comment]
Ok, thanks - I was trying to find a similar issue but didn't hit that one.

**Comment by alexcrichton** [comment]
Yeah I think this was fixed there, so closing.
